flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java

```diff
@@ -416,7 +452,18 @@ public int getUnannouncedCredit() {
 	 * @return Credit which was not announced to the sender yet.
 	 */
 	public int getAndResetUnannouncedCredit() {
-		return unannouncedCredit.getAndSet(0);
+		synchronized (bufferQueue) {
+			int credit = unannouncedCredit;
+			unannouncedCredit = 0;
+			if (credit == 0) {
```

Comment:
It is tricky to make zero credit has a special meaning, and we also break the previous assumption without negative credits. 

I think it might be better to define a separate message to describe the semantic of barrier alignment to unblock upstream side. Then the credit is always positive to be consistent as before, and we only define another blocked state on upstream side to control the buffer transport besides with positive credits.